### PR TITLE
New Package: py-xgboost

### PR DESCRIPTION
**Pull request by kean-smullen-nnl**
Creating a py-xgboost package for spack to install the python api for xgboost similar to the r package.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-xgboost/package.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    pypi = 'xgboost/xgboost-1.3.3.tar.gz'
+
+    version('1.3.3', sha256='397051647bb837915f3ff24afc7d49f7fca57630ffd00fb5ef66ae2a0881fb43')
+
+    variant('pandas',       default=False)
+    variant('scikit-learn', default=False)
+    variant('dask',         default=False)
+    variant('plotting',     default=False)
+
+    extends('python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xgboost/package.py]: This line isn't needed as it's already added by the base class

File: var/spack/repos/builtin/packages/py-xgboost/package.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    homepage  = 'https://xgboost.ai/'
+    pypi = 'xgboost/xgboost-1.3.3.tar.gz'
+
+    version('1.3.3', sha256='397051647bb837915f3ff24afc7d49f7fca57630ffd00fb5ef66ae2a0881fb43')
+
+    variant('pandas',       default=False)
+    variant('scikit-learn', default=False)
+    variant('dask',         default=False)
+    variant('plotting',     default=False)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xgboost/package.py]: These variants need descriptions

File: var/spack/repos/builtin/packages/py-xgboost/package.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    extends('python')
+    depends_on('cmake',         type='build')
+    depends_on('python@3.6:',   type=('build', 'run'))
+    depends_on('py-setuptools', type=('build'))
+    depends_on('py-numpy',      type=('build', 'run'))
+    depends_on('py-scipy',      type=('build', 'run'))
+
+    depends_on('py-pandas', when='+pandas', type=('build', 'run'))
+
+    depends_on('py-scikit-learn', when='+py-scikit-learn', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xgboost/package.py]:
```suggestion
    depends_on('py-scikit-learn', when='+scikit-learn', type=('build', 'run'))
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by adamjstewart** [comment]
Looks like the `setup.py` has some interesting options:
```python
# Options only effect `python setup.py install`, building `bdist_wheel`         
# requires using CMake directly.                                                
USER_OPTIONS = {                                                                
    # libxgboost options.                                                       
    'use-openmp': (None, 'Build with OpenMP support.', 1),                      
    'use-cuda':   (None, 'Build with GPU acceleration.', 0),                    
    'use-nccl':   (None, 'Build with NCCL to enable distributed GPU support.', 0),
    'build-with-shared-nccl': (None, 'Build with shared NCCL library.', 0),     
    'hide-cxx-symbols':       (None, 'Hide all C++ symbols during build.', 1),  
    'use-hdfs':   (None, 'Build with HDFS support', 0),                         
    'use-azure':  (None, 'Build with AZURE support.', 0),                       
    'use-s3':     (None, 'Build with S3 support', 0),                           
    'plugin-lz4': (None, 'Build lz4 plugin.', 0),                               
    'plugin-dense-parser': (None, 'Build dense parser plugin.', 0),             
    # Python specific                                                           
    'use-system-libxgboost': (None, 'Use libxgboost.so in system path.', 0)     
}                          
```
For example, we may be able to add `--use-system-libxgboost` and add a dependency on `xgboost` so that we don't need to build it twice to get the Python API. These other flags could have variants as well. This is completely up to you, as I don't currently use `xgboost`. If I decide to use it in the future, I may add these myself.

**Comment by kean-smullen-nnl** [comment]
@adamjstewart I pushed a new commit to address your comments and change request to fix the py-scikit-learn dependency (good catch there haha).
As for the --use-system-xgboost flag, I hadn't configured it that way because I wanted to mimic the r-xgboost setup.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
It just so happens that I need to use `py-xgboost` for a project now, so I'll stress test the recipe and let you know if I needed to make any modifications. Thanks for adding this, and good timing!

**Pull request merged**